File: sum root to leaf nums.py
# // Time Complexity : O(n)
# // Space Complexity : O(h)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        
        def dfs(cur, num):
            
            if not cur:
                return 0
            
            # to get the path traversal
            num = num * 10 + cur.val
            if not cur.left and not cur.right:
                return num
            
            return dfs(cur.left, num) + dfs(cur.right, num)
        
        return dfs(root, 0)
        

File: SumRootToLeafNums.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        Stack<TreeNode> st = new Stack<>();
        Stack<Integer> numSt = new Stack<>();

        int sum = 0;
        int currNum = 0;

        while(root != null || !st.isEmpty()){
            while(root != null){
                currNum = currNum * 10 + root.val;
                st.push(root);
                numSt.push(currNum);
                root = root.left;
            }
            //pop
            root = st.pop();
            currNum = numSt.pop();

            //check leaf
            if(root.left == null && root.right == null)
                sum += currNum;

            root = root.right;
        }
        return sum;
    }
}